test: bug fix in transaction_tests

glozow opened the pull request:
This is a followup to #19698.

- There was a bug in the `ExcludeIndividualFlags` function which is fixed here.
- Fixing this bug also showed that there is a test that's supposed to fail (already existing in tx_invalid.json) in tx_valid.json, so I removed it. Other than that, the tests should all pass.
- Also implements a few suggestions I received offline: removing the `OP_1`s from the invalid tests (similar to https://github.com/bitcoin/bitcoin/commit/19db590d044efe7d474a16720e5b56e7b55db54c), comments, and style.
- A few other small fixes, like adding asserts, putting all the flags in `mapFlagNames`, better error messages

glozow committed:
[test] remove invalid test from tx_valid.json

This exact test is also already in tx_invalid.json#L29
It will also test with no-P2SH flags, so duplicating with
different flags is not necessary.

glozow committed:
[test] fix bug in ExcludeIndividualFlags

PR #19168 introduced this function but it always returns an empty vector.

glozow committed:
[test] minor improvements / followups

Add missing script verify flags to mapFlagNames.
iterate through mapFlagNames values instead of bits.

BOOST_CHECK_MESSAGE better reports which test failed exactly, whereas
BOOST_ERROR was just incrementing the error counter.

glozow committed:
[test] remove unnecessary OP_1s from invalid tests

Similar to 19db590d044efe7d474a16720e5b56e7b55db54c, which removed these
for the valid tests. Not removing ones that cause a false/empty stack
error because these tests should fail due to being invalid with CSV/CLTV

glozow commented:
@MarcoFalke thanks for catching the bug - this should fix it

MarcoFalke commented on src/test/data/tx_valid.json:
```diff
@@ -62,10 +62,6 @@
 
-["An invalid P2SH Transaction"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "HASH160 0x14 0x7a052c840ba73af26755de42cf01cc9e0a49fef0 EQUAL"]],
-"010000000100010000000000000000000000000000000000000000000000000000000000000000000009085768617420697320ffffffff010000000000000000015100000000", "P2SH,CLEANSTACK,WITNESS"],
```
I think the test is testing that an invalid p2sh tx is valid before p2sh activation?

glozow replied:
With #19698, transaction_tests will check that excluding the `P2SH` flag will cause this test to pass, so it would still be a duplicate test.

glozow replied:
Added this comment to the commit log^

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -158,7 +158,7 @@ std::vector<unsigned int> ExcludeIndividualFlags(unsigned int flags)
     for (unsigned int i = 0; i < mapFlagNames.size(); ++i) {
         const unsigned int flags_excluding_i = TrimFlags(flags & ~(1U << i));
-        if (flags != flags_excluding_i && std::find(flags_combos.begin(), flags_combos.end(), flags_excluding_i) != flags_combos.end()) {
+        if (flags != flags_excluding_i && std::find(flags_combos.begin(), flags_combos.end(), flags_excluding_i) == flags_combos.end()) {
```
Any reason to re-invent the wheel and avoid `std::set`?

glozow replied:
Nope, I could change it to set.

glozow replied:
I've changed it to use a `std::set`

glozow commented:
Cirrus failure is in feature_blockfilterindex_prune.py, which I'm pretty sure is unrelated

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -152,14 +152,16 @@ unsigned int FillFlags(unsigned int flags)
 {
-    std::vector<unsigned int> flags_combos;
+    std::set<unsigned int> flags_combos;
     for (unsigned int i = 0; i < mapFlagNames.size(); ++i) {
```
how is the map supposed to be kept up to date? Looks like it is missing SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_PUBKEYTYPE?

glozow replied:
Hm, is it possible to iterate through all the script verify flags? Could add an assert that they're all present in `mapFlagNames`

MarcoFalke replied:
Oh, my comment was just an example. There are more missing than just the one, IIRC

glozow replied:
should all be there now

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -152,14 +152,16 @@ unsigned int FillFlags(unsigned int flags)
     return flags;
```
unrelated nit: Could `Assert(IsValidFlagCombination(flags));`? (Needs #21553)

glozow replied:
There are places where we need to fill the flags as well as check that they're a valid combination

glozow replied:
I completely misread this for some reason 🤦 sorry about that
Just rebased on top of #21553 and added the asserts

jnewbery commented on src/test/transaction_tests.cpp:
```diff
@@ -149,17 +154,20 @@ unsigned int FillFlags(unsigned int flags)
 
-// Return valid flags that are all except one flag for each flag
-std::vector<unsigned int> ExcludeIndividualFlags(unsigned int flags)
+// Exclude each possible script verify flag from flags. Returns a list of these flag combinations
```
This returns a set, not a list.

You could copy the set into a vector to avoid changing the return type if you really felt like it.

```diff
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index 3d6ad20b85..235862ec7a 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -161,7 +161,7 @@ unsigned int FillFlags(unsigned int flags)
 // Exclude each possible script verify flag from flags. Returns a list of these flag combinations
 // that are valid and without duplicates. For example: if flags=1111 and there are 4 possible flags,
 // this should return the vector [0111, 1011, 1101, 1110].
-std::set<unsigned int> ExcludeIndividualFlags(unsigned int flags)
+std::vector<unsigned int> ExcludeIndividualFlags(unsigned int flags)
 {
     std::set<unsigned int> flags_combos;
     for (unsigned int i = 0; i < mapFlagNames.size(); ++i) {
@@ -170,7 +170,7 @@ std::set<unsigned int> ExcludeIndividualFlags(unsigned int flags)
             flags_combos.insert(flags_excluding_i);
         }
     }
-    return flags_combos;
+    return {flags_combos.begin(), flags_combos.end()};
 }
 
 BOOST_FIXTURE_TEST_SUITE(transaction_tests, BasicTestingSetup)
```

glozow replied:
True, although hopefully it's obvious - will update if I need to retouch. Thanks!

glozow replied:
Fixed

jnewbery reviewed (commented):
utACK a944a21be41d1e5469822b1858c304c983010880

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -149,17 +154,20 @@ unsigned int FillFlags(unsigned int flags)
-    std::vector<unsigned int> flags_combos;
+    std::set<unsigned int> flags_combos;
     for (unsigned int i = 0; i < mapFlagNames.size(); ++i) {
         const unsigned int flags_excluding_i = TrimFlags(flags & ~(1U << i));
```
I still don't think this is right. A std::map doesn't order its elements based on the order they were inserted. 

Even if it did, the values aren't guaranteed to never miss a power of two.

glozow replied:
Sorry, what do you mean? The map `mapFlagNames` is just to identify the flags from strings in the json files. I guess if there wasn't a script verify flag for `1 << i` for some `i`, it would get ignored by the interpreter if we added it. So as long as we have the leftmost flag we should be okay?

MarcoFalke replied:
Let's assume mapFlagNames is only a single entry `{"EXAMPLE", 1U << 4}` (because for some reason other flags are no longer needed and have been removed).

The loop will only cover `1U << 1`, a non-existent flag.

glozow replied:
Ah true. Is there a way of checking whether an unsigned int corresponds to a script verify flag?

MarcoFalke replied:
Just iterate over the map and use the value of the key-value pair?

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -314,11 +321,15 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
+
             // Backwards compatibility of script verification flags: Adding any flag(s) should not validate an invalid transaction
             for (size_t i = 0; i < mapFlagNames.size(); i++) {
                 unsigned int flags = FillFlags(verify_flags | (1U << i));
```
same (and everywhere else)

glozow committed:
[test] check that mapFlagNames is up to date

There is no way to iterate through all script verification flags, and
it's not guaranteed that every power of 2 is used. Just make sure that
all flags in STANDARD_SCRIPT_VERIFY_FLAGS are present in mapFlagNames;
this covers all consensus and policy flags. If mapFlagNames has more
flags than STANDARD_SCRIPT_VERIFY_FLAGS, that's okay. Nonexistent flags
will be caught by the compiler.

glozow commented:
Updates:
Iterate through the values in `mapFlagNames` instead of all bits from 1 to size().
Check that all flags in `STANDARD_SCRIPT_VERIFY_FLAGS` are present in `mapFlagNames`. This would cover all relevant flags used for consensus and policy checks.

jnewbery commented on src/test/transaction_tests.cpp:
```diff
@@ -228,16 +248,15 @@ BOOST_AUTO_TEST_CASE(tx_valid)
+                unsigned int flags = TrimFlags(~(verify_flags | flag));
                 if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true)) {
-                    BOOST_ERROR("Tx unexpectedly failed with flag " << ToString(i) << " unset: " << strTest);
+                    BOOST_ERROR("Tx unexpectedly failed with flag " << ToString(flag) << " unset: " << strTest);
```
This doesn't need `ToString()`:

```suggestion
                    BOOST_ERROR("Tx unexpectedly failed with flag " << flag << " unset: " << strTest);
```

or alternatively, print the name of the flag:

```suggestion
                    BOOST_ERROR("Tx unexpectedly failed with flag " << _ << " unset: " << strTest);
```

(but probably change `_` to `flag_name` or similar)

glozow replied:
printing name of flag! :D

jnewbery commented on src/test/transaction_tests.cpp:
```diff
@@ -314,17 +333,21 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
             // Backwards compatibility of script verification flags: Adding any flag(s) should not validate an invalid transaction
-            for (size_t i = 0; i < mapFlagNames.size(); i++) {
-                unsigned int flags = FillFlags(verify_flags | (1U << i));
+            for (const auto & [_, flag] : mapFlagNames) {
```
```suggestion
            for (const auto& [_, flag] : mapFlagNames) {
```

jnewbery commented on src/test/transaction_tests.cpp:
```diff
@@ -314,17 +333,21 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
                 // Adding individual flags
                 if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ false)) {
-                    BOOST_ERROR("Tx unexpectedly passed with flag " << ToString(i) << " set: " << strTest);
+                    BOOST_ERROR("Tx unexpectedly passed with flag " << ToString(flag) << " set: " << strTest);
```
Again, `ToString()` is not necessary.

jnewbery reviewed (commented):
ACK 24f2951e11d0aa88e3b45271ec457abaadc6fde0

A few small style suggestions inline if you retouch this.

glozow commented:
Addressed the compiler warning about unused variable and @jnewbery's comments - thanks!

jnewbery commented:
utACK b109bde46a8730afbc09c107b802a2c321293f4b

MarcoFalke commented on src/test/data/tx_invalid.json:
```diff
@@ -124,15 +124,15 @@
+[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "500000001 CHECKLOCKTIMEVERIFY"]],
 "01000000010001000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000065cd1d", "CHECKLOCKTIMEVERIFY"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4294967295 CHECKLOCKTIMEVERIFY 1"]],
+[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4294967295 CHECKLOCKTIMEVERIFY"]],
```
It might be nice to check the exact script reject reason instead of just failure
